### ISSUE TITLE
Create a zap contract

### ISSUE THREAD

**Issue by ASlowCheetah** [pangolin-defi/exchange-contracts#3]
**Acceptance Criteria:**
* It needs to get permission to transfer the assets in behalf of the user
* It needs to have any further necessary permission for the contracts it's dealing with
* It only zaps from PNG
* It takes a vault as parameter, reads the tokens, the amoutns and swaps the PNG using Pangolin pairs for volume incentive
* It deposits the underlying assets to the vault
* It charges back any left over

**Comment by benber86** [comment]
How about renaming this to something other than Zapper (Swapper, VaultProxy?) and have it work as a proxy for all vaults in the way you suggest.

Let's say each Vault can be defined by a basic struct (for the sake of example here, I should actually think about a proper way to define this and post it under the Vault issue)

`
enum Numeraire {USD, AVAX}

    struct Vault {
        address[] assets;
        uint256[] weights;
        Numeraire numeraire;
    }
`

The numeraire is the asset we'll use to compute the value of the fund, and therefore also for rebalancing. We can't use PNG there because we would need PNG-denominated Chainlink feeds and right now CL only has USD feeds on Mainnet. There might be AVAX or ETH in the future, but there's no real use for a PNG-denominated feed outside of this project. And paying or running our own nodes for such a feed sounds overkill.

To me, it makes more sense to have the deposit asset be the same as the numeraire. It's less confusing from the perspective of users. 

Now of course the main counterpoint is liquidity, since the PNG pools have deeper liquidity than most others due to the incentivization mechanisms on Pangolin. But PNG could act as intermediary assets for all the swaps. This benefits the token as it creates volume and extra fees for LPs?

**Comment by ASlowCheetah** [comment]
> The numeraire is the asset we'll use to compute the value of the fund, and therefore also for rebalancing. We can't use PNG there because we would need PNG-denominated Chainlink feeds and right now CL only has USD feeds on Mainnet. There will likely be AVAX and ETH feeds in the future, but there's no real use for a PNG-denominated feed outside of this project. And paying or running our own nodes for such a feed sounds overkill.

You're absolutely right, and I think we might even accept USD or AVAX, but we should focus on having PNG as a first choice, in the end of the day, if someone supply USD we might be forced to swap that to AVAX or PNG to reduce the slippage risk

**Comment by benber86** [comment]
> You're absolutely right, and I think we might even accept USD or AVAX, but we should focus on having PNG as a first choice, in the end of the day, if someone supply USD we might be forced to swap that to AVAX or PNG to reduce the slippage risk.

Let's make it generalizable from the get go by having one `payable` deposit function for AVAX and another deposit function that takes `uint256 amount, address tokenERC20`  for any  ERC-20. The front-end could be set to call it with PNG if we want to force users down that road (I still think a USD entry would be more user friendly). 

We add a require `isWhiteListed` to the ERC-20 deposit function that checks whether the token being transferred is authorized.  And we have a few `ownerOnly` functions to handle the whitelist, which I guess would be PNG, USDT or DAI to start? 

> Later on we can upgrade this specific contract to have a higher control over the swap operation with better strategies for high liquidity, splitting into separated LPs for better swaps, etc...

How about having separate modules for handling the swaps? Because we're going to be doing swaps when we rebalance the Vault, so we want to be able to use those functionalities there too. 
In terms of architecture I suggest we have an entry point as "VaultEntryProxy"  following a <a href='https://blog.openzeppelin.com/proxy-patterns/'>proxy pattern</a>, which just points to the `VaultEntry` contract. This will let us upgrade the `VaultEntry` contract in the future if we need to.
Then `VaultEntry` queries the target Vault for its tokensets (struct as defined above), call the swapping logic, deposits the funds, and returns shares to the user.

**Comment by ASlowCheetah** [comment]
Sounds like a good plan, and +1 on the proxy, we will be building some proxies along the way as we need to have many parts that can be easily swapped/upgraded.

For the swaps, do we want to delegate it to a third contract or have it as a library? It would be good if we could avoid the delegateCall, but in this scenario it feels like it's worth the gas cost to have the flexibility
